src/net/download.rs

```diff
@@ -0,0 +1,1904 @@
+            let port = match url.port_or_known_default() {
+                Some(p) => p,
+                None => {
+                    warn!("Unsupported URL {:?}: unknown port", &url);
+                    continue;
+                }
+            };
+            match url.host() {
+                Some(url::Host::Domain(domain)) => {
+                    dns_client.queue_lookup(domain.clone(), port, get_epoch_time_ms() + self.dns_timeout)?;
+                    self.dns_lookups.insert(url_str.clone(), None);
+                    self.parsed_urls.insert(url_str, DNSRequest::new(domain.to_string(), port, 0));
+                },
+                Some(url::Host::Ipv4(addr)) => {
+                    self.dns_lookups.insert(url_str, Some(vec![SocketAddr::new(IpAddr::V4(addr), port)]));
+                }
+                Some(url::Host::Ipv6(addr)) => {
+                    self.dns_lookups.insert(url_str, Some(vec![SocketAddr::new(IpAddr::V6(addr), port)]));
+                },
+                None => {
+                    warn!("Unsupported URL {:?}", &url_str);
+                }
+            }
+        }
+
+        self.state = BlockDownloaderState::DNSLookupFinish;
+        Ok(())
+    }
+
+    pub fn dns_lookups_try_finish(&mut self, dns_client: &mut DNSClient) -> Result<bool, net_error> {
+        dns_client.try_recv()?;
+
+        let mut inflight = 0;
+        for (url_str, request) in self.parsed_urls.iter() {
+            match dns_client.poll_lookup(&request.host, request.port) {
+                Ok(Some(query_result)) => {
+                    if let Some(dns_result) = self.dns_lookups.get_mut(url_str) {
+                        // solicited
+                        match query_result.result {
+                            Ok(addrs) => {
+                                *dns_result = Some(addrs);
+                            },
+                            Err(msg) => {
+                                warn!("DNS failed to look up {:?}: {}", &url_str, msg);
+                            }
+                        }
+                    }
+                },
+                Ok(None) => {
+                    inflight += 1;
+                }
+                Err(e) => {
+                    warn!("DNS lookup failed on {:?}: {:?}", url_str, &e);
+                }
+            }
+        }
+
+        if inflight == 0 {
+            // done with DNS
+            dns_client.clear_all_requests();
+            self.state = BlockDownloaderState::GetBlocksBegin;
+        }
+
+        Ok(inflight == 0)
+    }
+
+    pub fn getblocks_begin(&mut self, mut requests: HashMap<BlockRequestKey, usize>) -> () {
+        assert_eq!(self.state, BlockDownloaderState::GetBlocksBegin);
+
+        // don't touch blocks-to-try -- that's managed by the peer network directly.
+        self.getblock_requests.clear();
+        for (key, event_id) in requests.drain() {
```

Comment:
Can this just set `self.getblock_requests = requests` rather than clearing and then copying from one to the other?

Reply:
Yup; can also do this in `inv.rs` and `neighbors.rs`.

Reply:
Fixed in 85af7fa